org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java

```diff
@@ -34,7 +34,7 @@ public URI toUri(String stringUri) {
 	 */
 	public URI toEmfUri(java.net.URI netUri) {
 		String decoded = toDecodedString(netUri);
-		URI uri = URI.createURI(decoded, false);
+		URI uri = URI.createURI(decoded, true, URI.FRAGMENT_NONE);
```

Comment:
I am not sure if URI.FRAGMENT_NONE is correct for all uses cases, so maybe this override must be done in org.eclipse.xtext.ide.server.UriExtensions and not in this class. What do you think, @szarnekow ?

Reply:
The assumption that there's no fragment is probably not possible to guarantee. I'd rather expect a fragment to be allowed.

What I'm thinking: The netUri is already correctly encoded. Why do we decode and re-encode it at all?

Reply:
I do not know either, but from testing, it looks like encoding of java.net.URI is different than encoding of an EMF URI.
netUri.toString() returns something like "something%3A%2Fpath%2Fto%2Fres%23ource" while uri.toString() returns "something:/path/to/res%23ource".

Reply:
@szarnekow , regarding "The assumption that there's no fragment is probably not possible to guarantee", I think we can do that in `org.eclipse.xtext.ide.server.UriExtensions`, but not in `org.eclipse.xtext.util.UriExtensions`. So I have done that bit only in the `ide.server` package class. I think that is guaranteed because `#` has no special meaning in LSP, so no LSP client is going to send EMF fragments. Why should it?

I have updated the PR to reflect that.

Reply:
I run the test test_toUri_16 and put a breakpoint in the debugger, then called toString() on both variables.

Reply:
The tests seem to assert some weird contract.
The already encoded string is apparently supposed to be parsed to a URI that has *some* escape-sequences removed where *allowed* while other escape sequences are supposed to be retained where *necessary*.